sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/BeamZetaSqlCalcRel.java

```diff
@@ -0,0 +1,259 @@
+    @Nullable private final String condition;
+    private final Schema inputSchema;
+    private final Schema outputSchema;
+    private transient List<PreparedExpression> projectExps;
+    @Nullable private transient PreparedExpression conditionExp;
+
+    CalcFn(
+        List<String> projects,
+        @Nullable String condition,
+        Schema inputSchema,
+        Schema outputSchema) {
+      Preconditions.checkArgument(projects.size() == outputSchema.getFieldCount());
+      this.projects = ImmutableList.copyOf(projects);
+      this.condition = condition;
+      this.inputSchema = inputSchema;
+      this.outputSchema = outputSchema;
+    }
+
+    @Setup
+    public void setup() {
+      AnalyzerOptions options = SqlAnalyzer.initAnalyzerOptions();
+      for (Field field : inputSchema.getFields()) {
+        options.addExpressionColumn(
+            sanitize(field.getName()), ZetaSqlUtils.beamFieldTypeToZetaSqlType(field.getType()));
+      }
+
+      projectExps = new ArrayList<>();
+      for (String project : projects) {
+        PreparedExpression projectExp = new PreparedExpression(sanitize(project));
+        projectExp.prepare(options);
+        projectExps.add(projectExp);
+      }
+      if (condition != null) {
+        conditionExp = new PreparedExpression(sanitize(condition));
+        conditionExp.prepare(options);
+      }
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      Map<String, Value> columns = new HashMap<>();
+      Row row = c.element();
+      for (Field field : inputSchema.getFields()) {
+        columns.put(
+            sanitize(field.getName()),
+            ZetaSqlUtils.javaObjectToZetaSqlValue(row.getValue(field.getName()), field.getType()));
+      }
+
+      // TODO[BEAM-8630]: support parameters in expression evaluation
+      // The map is empty because parameters in the query string have already been substituted.
+      Map<String, Value> params = new HashMap<>();
+
+      if (conditionExp != null && !conditionExp.execute(columns, params).getBoolValue()) {
+        return;
+      }
+
+      Row.Builder output = Row.withSchema(outputSchema);
+      for (int i = 0; i < outputSchema.getFieldCount(); i++) {
+        // TODO[BEAM-8630]: performance optimization by bundling the gPRC calls
+        Value v = projectExps.get(i).execute(columns, params);
+        output.addValue(
+            ZetaSqlUtils.zetaSqlValueToJavaObject(v, outputSchema.getField(i).getType()));
+      }
+      c.output(output.build());
+    }
+
+    @Teardown
+    public void teardown() {
+      for (PreparedExpression projectExp : projectExps) {
+        projectExp.close();
+      }
+      if (conditionExp != null) {
+        conditionExp.close();
+      }
+    }
+
+    // Replaces "$" with "_" because "$" is not allowed in a valid ZetaSQL identifier
+    // (ZetaSQL identifier syntax: [A-Za-z_][A-Za-z_0-9]*)
+    // TODO[BEAM-8630]: check if this is sufficient and correct
```

Comment:
I believe the TODO is don't do this here, make the unparser return valid SQL?

Reply:
This is not only related to unparsing. The field names chosen by Calcite has "$" (and then they appear in expressions). Although I do agree fixing the problem in Calcite (if possible) is a better solution. Added comments to note this.

Reply:
Thanks for adding the clarification. My understanding is that this approach will result in any "$" being replaced, that includes ones in string constants. For example `$1 = "$value"` becomes `_1 = "_value"`. I'm fine with this going in, but we need to fix it very soon.

Reply:
Yes that's correct. Agree we do need to fix this soon.